rust/guards/tests/tests.rs

```diff
@@ -0,0 +1,189 @@
+        match self
+            .env
+            .query_call(
+                self.canister_id,
+                Principal::anonymous(),
+                "is_item_processed",
+                Encode!(&item).unwrap(),
+            )
+            .expect("failed to get value")
+        {
+            WasmResult::Reply(bytes) => Decode!(&bytes, Option<bool>).unwrap(),
+            WasmResult::Reject(e) => {
+                panic!("Failed to get value: {:?}", e);
+            }
+        }
+    }
+
+    pub fn set_non_processed_items(&self, values: &[&str]) {
+        let values = values.iter().map(|s| s.to_string()).collect::<Vec<_>>();
+        let result = self
+            .env
+            .update_call(
+                self.canister_id,
+                Principal::anonymous(),
+                "set_non_processed_items",
+                Encode!(&values).unwrap(),
+            )
+            .expect("failed to set non-processed items");
+        assert_matches!(result, WasmResult::Reply(_));
+    }
+
+    pub fn process_single_item_with_panicking_callback(
+        &self,
+        item: &str,
+        future_type: &FutureType,
+    ) {
+        let res = self
+            .env
+            .update_call(
+                self.canister_id,
+                Principal::anonymous(),
+                "process_single_item_with_panicking_callback",
+                Encode!(&item, &future_type).unwrap(),
+            )
+            .expect_err("process_single_item_with_panicking_callback should panic");
+        assert_eq!(res.code, ErrorCode::CanisterCalledTrap);
+        assert!(res.description.contains("panicking callback!"));
+    }
+
+    pub fn submit_process_single_item_with_panicking_callback(
+        &self,
+        item: &str,
+        future_type: &FutureType,
+    ) -> Result<RawMessageId, UserError> {
+        self.env.submit_call(
+            self.canister_id,
+            Principal::anonymous(),
+            "process_single_item_with_panicking_callback",
+            Encode!(&item, &future_type).unwrap(),
+        )
+    }
+
+    pub fn process_all_items_with_panicking_callback(
+        &self,
+        panicking_item: &str,
+        future_type: &FutureType,
+    ) {
+        let res = self
+            .env
+            .update_call(
+                self.canister_id,
+                Principal::anonymous(),
+                "process_all_items_with_panicking_callback",
+                Encode!(&panicking_item, &future_type).unwrap(),
+            )
+            .expect_err("process_all_items_with_panicking_callback should panic");
+        assert_eq!(res.code, ErrorCode::CanisterCalledTrap);
+        assert!(res.description.contains("panicking callback!"));
+    }
+}
+
+impl Default for CanisterSetup {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(CandidType, Debug, PartialEq, Eq)]
+pub enum FutureType {
+    TrueAsyncCall,
+    FalseAsyncCall,
+}
```

Comment:
Can this not be imported instead of copying its definition?

Reply:
I tried but this doesn't seem top work with the `crate-type` being `cdylib` (see https://users.rust-lang.org/t/does-crate-type-has-effect-on-integration-test/76634).  A solution would be to put the type in a separate crate that can be both imported in the canister and as a dev-dependency but this seems overkill here.

Reply:
I agree it's an overkill so we can keep things as they are.